src/samples/plugins/ModularSessionKeyPlugin.sol

```diff
@@ -0,0 +1,336 @@
+    {
+        bytes memory key = account.allocateAssociatedStorageKey(0, 1);
+        StoragePointer ptr = key.associatedStorageLookup(keccak256(abi.encodePacked(tempOwner, allowedSelector)));
+        SessionInfo storage sessionInfo = _castPtrToStruct(ptr);
+        _validAfter = sessionInfo.validAfter;
+        _validUntil = sessionInfo.validUntil;
+    }
+
+    // ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    // ┃    Plugin interface functions    ┃
+    // ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+    /// @inheritdoc BasePlugin
+    function onInstall(bytes calldata data) external override {
+        if (data.length != 0) {
+            (
+                address[] memory tempOwners,
+                bytes4[] memory allowedSelectors,
+                uint48[] memory validAfters,
+                uint48[] memory validUntils
+            ) = abi.decode(data, (address[], bytes4[], uint48[], uint48[]));
+            if (
+                tempOwners.length != allowedSelectors.length || tempOwners.length != validAfters.length
+                    || tempOwners.length != validUntils.length
+            ) {
+                revert WrongDataLength();
+            }
+            for (uint256 i = 0; i < tempOwners.length; i++) {
+                _addSessionKey(msg.sender, tempOwners[i], allowedSelectors[i], validAfters[i], validUntils[i]);
+            }
+        }
+    }
+
+    /// @inheritdoc BasePlugin
+    function onUninstall(bytes calldata data) external override {
+        if (data.length != 0) {
+            (address[] memory tempOwners, bytes4[] memory allowedSelectors) =
+                abi.decode(data, (address[], bytes4[]));
+            if (tempOwners.length != allowedSelectors.length) {
+                revert WrongDataLength();
+            }
+            for (uint256 i = 0; i < tempOwners.length; i++) {
+                _removeSessionKey(msg.sender, tempOwners[i], allowedSelectors[i]);
+            }
+        }
+    }
+
+    /// @inheritdoc BasePlugin
+    function userOpValidationFunction(uint8 functionId, UserOperation calldata userOp, bytes32 userOpHash)
+        external
+        view
+        override
+        returns (uint256)
+    {
+        if (functionId == uint8(FunctionId.USER_OP_VALIDATION_TEMPORARY_OWNER)) {
+            (address signer,) = userOpHash.toEthSignedMessageHash().tryRecover(userOp.signature);
+            bytes4 selector = bytes4(userOp.callData[0:4]);
+            bytes memory key = userOp.sender.allocateAssociatedStorageKey(0, 1);
+            StoragePointer ptr = key.associatedStorageLookup(keccak256(abi.encodePacked(signer, selector)));
+            SessionInfo storage duration = _castPtrToStruct(ptr);
+            uint48 validAfter = duration.validAfter;
+            uint48 validUntil = duration.validUntil;
+
+            if (validUntil != 0) {
+                return _packValidationData(false, validUntil, validAfter);
+            }
+            return _SIG_VALIDATION_FAILED;
```

Comment:
Delaying SIG_VALIDATION_FAILED like this helps with gas estimation from bundlers when using dummy signatures.

```suggestion
            return _packValidationData(validUntil == 0, validUntil, validAfter);
```